tests/cpptests/test_cpp_thpool.cpp

```diff
@@ -147,3 +147,60 @@ TEST_F(PriorityThpoolTestWithoutPrivilegedThreads, CombinationTest) {
     ASSERT_LT(arr[2], arr[4]);
     ASSERT_LT(arr[4], arr[3]);
 }
+
+/* ========================== NUM_THREADS = 2, NUM_PRIVILEGED = 0 ========================== */
+THPOOL_TEST_SUITE(PriorityThpoolTestFunctionality, 1, 0)
+
+void sleep_1_and_set(size_t *time_ms) {
+    std::this_thread::sleep_for(std::chrono::milliseconds(*time_ms));
+}
+
+TEST_P(PriorityThpoolTestFunctionality, TestTerminateWhenEmpty) {
+    // Pause the thread pool before adding tasks, to validate that jobs won't get executed before
+    // all other jobs are inserted into the queue.
+    redisearch_thpool_terminate_pause_threads(this->pool);
+
+    // Fill the job queue with tasks.
+    size_t time_ms = 1;
+    redisearch_thpool_add_work(this->pool, (void (*)(void *))sleep_1_and_set, &time_ms, THPOOL_PRIORITY_HIGH);
+
+    redisearch_thpool_resume_threads(this->pool);
+
+    redisearch_thpool_terminate_when_empty(this->pool);
+    redisearch_thpool_wait(this->pool);
+
+    thpool_stats stats = redisearch_thpool_get_stats(this->pool);
+    while (stats.num_threads_alive) {
+        usleep(1);
+        stats = redisearch_thpool_get_stats(this->pool);
+    }
+
+    // Recreate threads
+    redisearch_thpool_add_work(this->pool, (void (*)(void *))sleep_1_and_set, &time_ms, THPOOL_PRIORITY_HIGH);
+
+    stats = redisearch_thpool_get_stats(this->pool);
+    while (!stats.num_threads_alive) {
+        usleep(1);
+        stats = redisearch_thpool_get_stats(this->pool);
+    }
+
+}
+
+TEST_P(PriorityThpoolTestFunctionality, TestPauseResume) {
+    // add job long enough so they won't finish until we call pause
+    size_t time_ms = 1000;
+    redisearch_thpool_add_work(this->pool, (void (*)(void *))sleep_1_and_set, &time_ms, THPOOL_PRIORITY_HIGH);
+
+    // wait for them to be pulled
+    size_t curr_working  = redisearch_thpool_num_threads_working(this->pool);
+    while (curr_working != 1) {
+        usleep(1);
+    }
+
+    // pause threads
+    redisearch_thpool_terminate_pause_threads(this->pool);
```

Comment:
Not sure that I understand the purpose of this test - if you pause the threads after the single job is done, the `num_working=0` and `num_working =0` should exist anyway no, even without pausing, no? Did you mean to check `num_alive`? or am I missing something...

Reply:
I wait for it to be pulled, not done. I'm pausing the threadpool while it's in progress (i assume it won't be finished in 1 sec, see line 195 comment).
I added this test since in the first attempt I forgot to wait for the `num_jobs_in_progress` in the pause function, and it caused tests that expect a specific amount of jobs to be done, to be flaky.